src/compiler/scala/tools/nsc/PhaseAssembly.scala

```diff
@@ -112,16 +112,19 @@ trait PhaseAssembly {
 
       if (node.level < lvl) node.level = lvl
 
-      var hls = Nil ++ node.before.filter(_.hard)
-      while (hls.size > 0) {
-        for (hl <- hls) {
-          node.phaseobj = Some(node.phaseobj.get ++ hl.frm.phaseobj.get)
-          node.before = hl.frm.before
-          nodes -= hl.frm.phasename
-          edges -= hl
-          for (edge <- node.before) edge.to = node
+      var befores = node.before
+      def hasHardLinks() = befores.exists(_.hard)
+      while (hasHardLinks()) {
+        for (hl <- befores) {
+          if (hl.hard) {
+            node.phaseobj = Some(node.phaseobj.get ++ hl.frm.phaseobj.get)
+            node.before = hl.frm.before
+            nodes -= hl.frm.phasename
+            edges -= hl
+            for (edge <- node.before) edge.to = node
+          }
         }
-        hls = Nil ++ node.before.filter(_.hard)
```

Comment:
Not sure why this archaic idiom is so bad. Is before a large set of soft dependencies? Or making an intermediate one-element hashset just to iterator over it to build a List?

Reply:
At a micro-level, its wasteful because it the `Set.filter` creates an intermediate set. `node.before.iterator.filter(_.hard).toList` would be an improvement. But my refactoring was based on the belief that we could just remove the `filter` altogether and just filter out soft links in the lambda argument to `befores.foreach`.

Reply:
At least in 2.12, empty `mutable.HashSet`-s aren't that tiny -- they contain a 32-element array for the hash table plus some more data.

Reply:
Aha.